sql/src/main/java/io/crate/lucene/ArrayLengthQuery.java

```diff
@@ -0,0 +1,413 @@
+import io.crate.expression.operator.GteOperator;
+import io.crate.expression.operator.LtOperator;
+import io.crate.expression.operator.LteOperator;
+import io.crate.expression.operator.Operators;
+import io.crate.expression.symbol.Function;
+import io.crate.expression.symbol.Symbol;
+import io.crate.metadata.ColumnIdent;
+import io.crate.metadata.Reference;
+import io.crate.types.BooleanType;
+import io.crate.types.ByteType;
+import io.crate.types.CollectionType;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+import io.crate.types.DoubleType;
+import io.crate.types.FloatType;
+import io.crate.types.GeoPointType;
+import io.crate.types.IntegerType;
+import io.crate.types.IpType;
+import io.crate.types.LongType;
+import io.crate.types.ShortType;
+import io.crate.types.StringType;
+import io.crate.types.TimestampType;
+import org.apache.lucene.index.DocValues;
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.SortedNumericDocValues;
+import org.apache.lucene.index.SortedSetDocValues;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.ConstantScoreScorer;
+import org.apache.lucene.search.ConstantScoreWeight;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.TwoPhaseIterator;
+import org.apache.lucene.search.Weight;
+import org.elasticsearch.common.lucene.search.Queries;
+import org.elasticsearch.index.fielddata.FieldData;
+import org.elasticsearch.index.fielddata.SortedBinaryDocValues;
+import org.elasticsearch.index.mapper.MappedFieldType;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.List;
+import java.util.function.IntPredicate;
+import java.util.function.IntUnaryOperator;
+
+import static io.crate.lucene.LuceneQueryBuilder.genericFunctionFilter;
+
+public final class ArrayLengthQuery implements InnerFunctionToQuery {
+
+    /**
+     * <pre>
+     * {@code
+     *  array_length(arr, dim) > 0
+     *      |                  |
+     *    inner              parent
+     * }
+     * </pre>
+     */
+    @Nullable
+    @Override
+    public Query apply(Function parent, Function arrayLength, LuceneQueryBuilder.Context context) {
+        String parentName = parent.info().ident().name();
+        if (!Operators.COMPARISON_OPERATORS.contains(parentName)) {
+            return null;
+        }
+        List<Symbol> parentArgs = parent.arguments();
+        Symbol cmpSymbol = parentArgs.get(1);
+        if (!(cmpSymbol instanceof Input)) {
+            return null;
+        }
+        Number cmpNumber = (Number) ((Input) cmpSymbol).value();
+        if (cmpNumber == null) {
+            // TODO: this case is never reached due to early normalization + no-match detection
```

Comment:
shall we drop this code block then?

Reply:
I was wondering if I should add a unit test for this case so that it's hit or replace it with an assertion.

Reply:
yeah, an assertion would be great